### PR TITLE
Free nanopb mallocs when losing references to them

### LINKED ISSUE #11027: Multiple memory leaks in FirebaseSessions >>> FIRSESEncodeString
### Description

I am seeing multiple memory leaks in FirebaseSessions:

leaks

leaks_2

### Reproducing the issue

Calling

`[FIRApp configure];`

in `didFinishLaunchingWithOptions`

causes the leaks.

### Firebase SDK Version

10.7.0

### Xcode Version

14.3 (14E222b)

### Installation Method

CocoaPods

### Firebase Product(s)

Analytics, Crashlytics, Remote Config

### Targeted Platforms

iOS

### Relevant Log Output

_No response_

### If using Swift Package Manager, the project's Package.resolved

_No response_

### If using CocoaPods, the project's Podfile.lock

<details>
<summary>Expand <code>Podfile.lock</code> snippet</summary>
<br>

```yml

PODS:
  - Firebase/Analytics (10.7.0):
    - Firebase/Core
  - Firebase/Core (10.7.0):
    - Firebase/CoreOnly
    - FirebaseAnalytics (~> 10.7.0)
  - Firebase/CoreOnly (10.7.0):
    - FirebaseCore (= 10.7.0)
  - Firebase/Crashlytics (10.7.0):
    - Firebase/CoreOnly
    - FirebaseCrashlytics (~> 10.7.0)
  - Firebase/RemoteConfig (10.7.0):
    - Firebase/CoreOnly
    - FirebaseRemoteConfig (~> 10.7.0)
  - FirebaseABTesting (10.7.0):
    - FirebaseCore (~> 10.0)
  - FirebaseAnalytics (10.7.0):
    - FirebaseAnalytics/AdIdSupport (= 10.7.0)
    - FirebaseCore (~> 10.0)
    - FirebaseInstallations (~> 10.0)
    - GoogleUtilities/AppDelegateSwizzler (~> 7.8)
    - GoogleUtilities/MethodSwizzler (~> 7.8)
    - GoogleUtilities/Network (~> 7.8)
    - "GoogleUtilities/NSData+zlib (~> 7.8)"
    - nanopb (< 2.30910.0, >= 2.30908.0)
  - FirebaseAnalytics/AdIdSupport (10.7.0):
    - FirebaseCore (~> 10.0)
    - FirebaseInstallations (~> 10.0)
    - GoogleAppMeasurement (= 10.7.0)
    - GoogleUtilities/AppDelegateSwizzler (~> 7.8)
    - GoogleUtilities/MethodSwizzler (~> 7.8)
    - GoogleUtilities/Network (~> 7.8)
    - "GoogleUtilities/NSData+zlib (~> 7.8)"
    - nanopb (< 2.30910.0, >= 2.30908.0)
  - FirebaseCore (10.7.0):
    - FirebaseCoreInternal (~> 10.0)
    - GoogleUtilities/Environment (~> 7.8)
    - GoogleUtilities/Logger (~> 7.8)
  - FirebaseCoreExtension (10.7.0):
    - FirebaseCore (~> 10.0)
  - FirebaseCoreInternal (10.7.0):
    - "GoogleUtilities/NSData+zlib (~> 7.8)"
  - FirebaseCrashlytics (10.7.0):
    - FirebaseCore (~> 10.5)
    - FirebaseInstallations (~> 10.0)
    - FirebaseSessions (~> 10.5)
    - GoogleDataTransport (~> 9.2)
    - GoogleUtilities/Environment (~> 7.8)
    - nanopb (< 2.30910.0, >= 2.30908.0)
    - PromisesObjC (~> 2.1)
  - FirebaseInstallations (10.7.0):
    - FirebaseCore (~> 10.0)
    - GoogleUtilities/Environment (~> 7.8)
    - GoogleUtilities/UserDefaults (~> 7.8)
    - PromisesObjC (~> 2.1)
  - FirebaseRemoteConfig (10.7.0):
    - FirebaseABTesting (~> 10.0)
    - FirebaseCore (~> 10.0)
    - FirebaseInstallations (~> 10.0)
    - GoogleUtilities/Environment (~> 7.8)
    - "GoogleUtilities/NSData+zlib (~> 7.8)"
  - FirebaseSessions (10.7.0):
    - FirebaseCore (~> 10.5)
    - FirebaseCoreExtension (~> 10.0)
    - FirebaseInstallations (~> 10.0)
    - GoogleDataTransport (~> 9.2)
    - GoogleUtilities/Environment (~> 7.10)
    - nanopb (< 2.30910.0, >= 2.30908.0)
    - PromisesSwift (~> 2.1)
  - GoogleAppMeasurement (10.7.0):
    - GoogleAppMeasurement/AdIdSupport (= 10.7.0)
    - GoogleUtilities/AppDelegateSwizzler (~> 7.8)
    - GoogleUtilities/MethodSwizzler (~> 7.8)
    - GoogleUtilities/Network (~> 7.8)
    - "GoogleUtilities/NSData+zlib (~> 7.8)"
    - nanopb (< 2.30910.0, >= 2.30908.0)
  - GoogleAppMeasurement/AdIdSupport (10.7.0):
    - GoogleAppMeasurement/WithoutAdIdSupport (= 10.7.0)
    - GoogleUtilities/AppDelegateSwizzler (~> 7.8)
    - GoogleUtilities/MethodSwizzler (~> 7.8)
    - GoogleUtilities/Network (~> 7.8)
    - "GoogleUtilities/NSData+zlib (~> 7.8)"
    - nanopb (< 2.30910.0, >= 2.30908.0)
  - GoogleAppMeasurement/WithoutAdIdSupport (10.7.0):
    - GoogleUtilities/AppDelegateSwizzler (~> 7.8)
    - GoogleUtilities/MethodSwizzler (~> 7.8)
    - GoogleUtilities/Network (~> 7.8)
    - "GoogleUtilities/NSData+zlib (~> 7.8)"
    - nanopb (< 2.30910.0, >= 2.30908.0)
  - GoogleDataTransport (9.2.1):
    - GoogleUtilities/Environment (~> 7.7)
    - nanopb (< 2.30910.0, >= 2.30908.0)
    - PromisesObjC (< 3.0, >= 1.2)
  - GoogleUtilities/AppDelegateSwizzler (7.11.0):
    - GoogleUtilities/Environment
    - GoogleUtilities/Logger
    - GoogleUtilities/Network
  - GoogleUtilities/Environment (7.11.0):
    - PromisesObjC (< 3.0, >= 1.2)
  - GoogleUtilities/Logger (7.11.0):
    - GoogleUtilities/Environment
  - GoogleUtilities/MethodSwizzler (7.11.0):
    - GoogleUtilities/Logger
  - GoogleUtilities/Network (7.11.0):
    - GoogleUtilities/Logger
    - "GoogleUtilities/NSData+zlib"
    - GoogleUtilities/Reachability
  - "GoogleUtilities/NSData+zlib (7.11.0)"
  - GoogleUtilities/Reachability (7.11.0):
    - GoogleUtilities/Logger
  - GoogleUtilities/UserDefaults (7.11.0):
    - GoogleUtilities/Logger
  - nanopb (2.30909.0):
    - nanopb/decode (= 2.30909.0)
    - nanopb/encode (= 2.30909.0)
  - nanopb/decode (2.30909.0)
  - nanopb/encode (2.30909.0)
  - PromisesObjC (2.2.0)
  - PromisesSwift (2.2.0):
    - PromisesObjC (= 2.2.0)

DEPENDENCIES:
  - Firebase/Analytics
  - Firebase/Crashlytics
  - Firebase/RemoteConfig

SPEC REPOS:
  trunk:
    - Firebase
    - FirebaseABTesting
    - FirebaseAnalytics
    - FirebaseCore
    - FirebaseCoreExtension
    - FirebaseCoreInternal
    - FirebaseCrashlytics
    - FirebaseInstallations
    - FirebaseRemoteConfig
    - FirebaseSessions
    - GoogleAppMeasurement
    - GoogleDataTransport
    - GoogleUtilities
    - nanopb
    - PromisesObjC
    - PromisesSwift


SPEC CHECKSUMS:
  Firebase: 0219acf760880eeec8ce479895bd7767466d9f81
  FirebaseABTesting: 76c8297fd026074e0366dc941d265d1be80a56d5
  FirebaseAnalytics: f8133442ee6f8512e28ff19e62ce15398bfaeace
  FirebaseCore: e317665b9d744727a97e623edbbed009320afdd7
  FirebaseCoreExtension: f17247ba8c61e4d3c8d136b5e2de3cb4ac6a85b6
  FirebaseCoreInternal: 8845798510aae74703467480f71ac613788d0696
  FirebaseCrashlytics: 35fdd1a433b31e28adcf5c8933f4c526691a1e0b
  FirebaseInstallations: 59c0e4c7a816a0f76710d83f77e5369b3e45eb96
  FirebaseRemoteConfig: d5de62211e2eaa2152d8ee85a23c301b70887a74
  FirebaseSessions: 34e5c084da010ef3802cbc062b822e513c9e6318
  GoogleAppMeasurement: fe17c92a32207dd5cdd4e8d742767f2da74857f6
  GoogleDataTransport: ea169759df570f4e37bdee1623ec32a7e64e67c4
  GoogleUtilities: c2bdc4cf2ce786c4d2e6b3bcfd599a25ca78f06f
  nanopb: b552cce312b6c8484180ef47159bc0f65a1f0431
  PromisesObjC: 09985d6d70fbe7878040aa746d78236e6946d2ef
  PromisesSwift: cf9eb58666a43bbe007302226e510b16c1e10959

PODFILE CHECKSUM: 233734da80802916263187863a831982228040cf

COCOAPODS: 1.12.0

```

</details>

### PR DESCRIPTION
**Pull request by morganchen12**
Fixes #11027.

### CODE REVIEW

File: FirebaseSessions/Sources/SessionStartEvent.swift [hunk 1]
```diff
@@ -74,8 +74,25 @@ class SessionStartEvent: NSObject, GDTCOREventDataObject {
...
+    ]
+    for pointer in garbage {
+      nanopb_free(pointer)
+    }
+  }
+
   func setInstallationID(installationId: String) {
+    let oldID = proto.session_data.firebase_installation_id
     proto.session_data.firebase_installation_id = makeProtoString(installationId)
+    nanopb_free(oldID)
```
Comment (ncooke3) [inline FirebaseSessions/Sources/SessionStartEvent.swift:95]: what does these added lines do?

File: FirebaseSessions/SourcesObjC/NanoPB/FIRSESNanoPBHelpers.m [hunk 2]
```diff
@@ -86,7 +94,7 @@
   if (pbBytes == NULL) {
     return NULL;
   }
-  memcpy(pbBytes->bytes, [data bytes], data.length);
+  [data getBytes:pbBytes->bytes length:data.length];
```
Comment (ncooke3) [inline FirebaseSessions/SourcesObjC/NanoPB/FIRSESNanoPBHelpers.m:97]: also here, why the change?

**Review by ncooke3 — COMMENTED** [review]
nice @morganchen12–– I left some questions for my own learning.

Also, optional, but anyway reasonable way to add a unit test? Maybe some loop that previously would result in an OOM but won't with your changes? Not critical though if it;s infeasible.

File: FirebaseSessions/Sources/SessionStartEvent.swift (hunk 1 above)
Comment (morganchen12) [reply FirebaseSessions/Sources/SessionStartEvent.swift:95]: Some of the fields in the nanopb proto are `malloc`ed byte buffers instead of integer, struct, or enum types, so they need to be manually freed.

File: FirebaseSessions/SourcesObjC/NanoPB/FIRSESNanoPBHelpers.m (hunk 2 above)
Comment (morganchen12) [reply FirebaseSessions/SourcesObjC/NanoPB/FIRSESNanoPBHelpers.m:97]: This change isn't necessary; I can remove it. I made it while monkeying around in these files and it adheres to the principle of use the highest level of abstraction available so I left it.

Comment (ncooke3) [reply FirebaseSessions/SourcesObjC/NanoPB/FIRSESNanoPBHelpers.m:97]: Got it– I thought it was necessary for the fix. Keeping it SGTM if you prefer.

File: FirebaseSessions/Sources/SessionStartEvent.swift (hunk 1 above)
Comment (morganchen12) [reply FirebaseSessions/Sources/SessionStartEvent.swift:95]: I'm keeping a reference to it around so I can pass it to the `nanopb_free` function. These are all `UnsafeMutablePointer`s, so they're references and the string buffers they point to aren't copied when assigning to a new variable.

I was worried something like
```swift
nanopb_free(proto.session_data.firebase_installation_id)
```
could cause issues with threaded access since the value held by the variable would be briefly garbage, though this implementation isn't thread-safe either.

**Review by ncooke3 — APPROVED** [review]

**Comment by morganchen12** [comment]
Tested manually via the quickstart and there's no more leaks.

**Review by visumickey — COMMENTED** [review]
Just have a comment about the oldId and freeing oldId technique.

File: FirebaseSessions/Sources/SessionStartEvent.swift [hunk 3]
```diff
@@ -74,8 +74,28 @@ class SessionStartEvent: NSObject, GDTCOREventDataObject {
...
+      proto.application_info.apple_app_info.bundle_short_version,
+      proto.application_info.apple_app_info.mcc_mnc,
+    ]
+    for pointer in garbage {
+      nanopb_free(pointer)
+    }
+  }
+
   func setInstallationID(installationId: String) {
+    let oldID = proto.session_data.firebase_installation_id
```
[the remark this answers is not in the document]
Comment (visumickey) [reply FirebaseSessions/Sources/SessionStartEvent.swift:96]: An example of how we did this in Fireperf. https://github.com/firebase/firebase-ios-sdk/blob/master/FirebasePerformance/Sources/FPRNanoPbUtils.m#L246

[the remark this answers is not in the document]
Comment (morganchen12) [reply FirebaseSessions/Sources/SessionStartEvent.swift:96]: As it currently is used, this change doesn't prevent any memory leaks, and is unnecessary if the API contract stipulates this field can only be set once for the lifetime of the session event object.

I'm not sure how changing the initialization method makes this leak go away: the allocation happens in the call to `makeProtoString` below.

File: FirebaseSessions/Sources/SessionStartEvent.swift [hunk 4]
```diff
@@ -74,8 +74,28 @@ class SessionStartEvent: NSObject, GDTCOREventDataObject {
       .performance = firebase_appquality_sessions_DataCollectionState_COLLECTION_SDK_NOT_INSTALLED
   }
 
+  deinit {
```
[the remark this answers is not in the document]
Comment (morganchen12) [reply FirebaseSessions/Sources/SessionStartEvent.swift:77]: Yes, this approach is definitely error-prone. If the proto definition is ever regenerated with new fields that need to be freed, we won't be notified and this method will start leaking memory.

[the remark this answers is not in the document]
Comment (morganchen12) [reply FirebaseSessions/Sources/SessionStartEvent.swift:77]: I talked with @ncooke3 and we'll merge this as-is to make the release deadline, but I'll address feedback in a follow-up change. As it currently stands this change could definitely use some improvement.

**Comment by morganchen12** [comment]
Resolving comments and addressing feedback in a follow-up to make the release deadline.

File: FirebaseSessions/Sources/SessionStartEvent.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (morganchen12) [reply FirebaseSessions/Sources/SessionStartEvent.swift:77]: @visumickey can you elaborate on what you mean by the parent message?

[the remark this answers is not in the document]
Comment (visumickey) [reply FirebaseSessions/Sources/SessionStartEvent.swift:77]: My parent message message I meant the top level message. Example: If we deallocate the top-level message, would that automatically reallocate the sub-top level message?

[the remark this answers is not in the document]
Comment (morganchen12) [reply FirebaseSessions/Sources/SessionStartEvent.swift:77]: Hey, sorry for the late reply. I couldn't find a deinit for the nanopb-generated struct type that owns these pointers.
